### PR TITLE
protobuf: cache temp buffers

### PR DESCRIPTION
**Pull request by carl-mastrangelo**
Alternate implementation of https://github.com/grpc/grpc-java/pull/2288

### CODE REVIEW

**Comment by carl-mastrangelo** [comment]
Setting the transport benchmark to use megabyte messages sees the following results:

```
Before:
TransportBenchmark.unaryCall1024                                       true        NETTY  sample  4564   2188854.745 ±  71456.423   ns/op
TransportBenchmark.unaryCall1024:unaryCall1024·p0.00                   true        NETTY  sample         1875968.000                ns/op
TransportBenchmark.unaryCall1024:unaryCall1024·p0.50                   true        NETTY  sample         2105344.000                ns/op
TransportBenchmark.unaryCall1024:unaryCall1024·p0.90                   true        NETTY  sample         2396160.000                ns/op
TransportBenchmark.unaryCall1024:unaryCall1024·p0.95                   true        NETTY  sample         2535424.000                ns/op
TransportBenchmark.unaryCall1024:unaryCall1024·p0.99                   true        NETTY  sample         3011993.600                ns/op
TransportBenchmark.unaryCall1024:unaryCall1024·p0.999                  true        NETTY  sample         7471595.520                ns/op
TransportBenchmark.unaryCall1024:unaryCall1024·p0.9999                 true        NETTY  sample        99090432.000                ns/op
TransportBenchmark.unaryCall1024:unaryCall1024·p1.00                   true        NETTY  sample        99090432.000                ns/op
TransportBenchmark.unaryCall1024:·gc.alloc.rate                        true        NETTY  sample    10      2787.784 ±    169.945  MB/sec
TransportBenchmark.unaryCall1024:·gc.alloc.rate.norm                   true        NETTY  sample    10   6415272.837 ±    262.046    B/op
TransportBenchmark.unaryCall1024:·gc.churn.PS_Eden_Space               true        NETTY  sample    10      2815.863 ±    429.465  MB/sec
TransportBenchmark.unaryCall1024:·gc.churn.PS_Eden_Space.norm          true        NETTY  sample    10   6483440.294 ± 947355.959    B/op
TransportBenchmark.unaryCall1024:·gc.churn.PS_Survivor_Space           true        NETTY  sample    10         2.143 ±      1.623  MB/sec
TransportBenchmark.unaryCall1024:·gc.churn.PS_Survivor_Space.norm      true        NETTY  sample    10      4873.798 ±   3679.598    B/op
TransportBenchmark.unaryCall1024:·gc.count                             true        NETTY  sample    10        42.000               counts
TransportBenchmark.unaryCall1024:·gc.time                              true        NETTY  sample    10       155.000                   ms


After:
Benchmark                                                          (direct)  (transport)    Mode   Cnt         Score        Error   Units
TransportBenchmark.unaryCall1024                                       true        NETTY  sample  5037   1982881.569 ±  16738.841   ns/op
TransportBenchmark.unaryCall1024:unaryCall1024·p0.00                   true        NETTY  sample         1683456.000                ns/op
TransportBenchmark.unaryCall1024:unaryCall1024·p0.50                   true        NETTY  sample         1918976.000                ns/op
TransportBenchmark.unaryCall1024:unaryCall1024·p0.90                   true        NETTY  sample         2232320.000                ns/op
TransportBenchmark.unaryCall1024:unaryCall1024·p0.95                   true        NETTY  sample         2330624.000                ns/op
TransportBenchmark.unaryCall1024:unaryCall1024·p0.99                   true        NETTY  sample         2729574.400                ns/op
TransportBenchmark.unaryCall1024:unaryCall1024·p0.999                  true        NETTY  sample         6127304.704                ns/op
TransportBenchmark.unaryCall1024:unaryCall1024·p0.9999                 true        NETTY  sample        15515648.000                ns/op
TransportBenchmark.unaryCall1024:unaryCall1024·p1.00                   true        NETTY  sample        15515648.000                ns/op
TransportBenchmark.unaryCall1024:·gc.alloc.rate                        true        NETTY  sample    10      2071.435 ±    141.669  MB/sec
TransportBenchmark.unaryCall1024:·gc.alloc.rate.norm                   true        NETTY  sample    10   4318096.849 ±    269.655    B/op
TransportBenchmark.unaryCall1024:·gc.churn.PS_Eden_Space               true        NETTY  sample    10      2076.282 ±    323.504  MB/sec
TransportBenchmark.unaryCall1024:·gc.churn.PS_Eden_Space.norm          true        NETTY  sample    10   4335884.918 ± 729189.378    B/op
TransportBenchmark.unaryCall1024:·gc.churn.PS_Survivor_Space           true        NETTY  sample    10         1.567 ±      1.238  MB/sec
TransportBenchmark.unaryCall1024:·gc.churn.PS_Survivor_Space.norm      true        NETTY  sample    10      3274.883 ±   2640.345    B/op
TransportBenchmark.unaryCall1024:·gc.count                             true        NETTY  sample    10        31.000               counts
TransportBenchmark.unaryCall1024:·gc.time                              true        NETTY  sample    10        51.000                   ms

```

**Comment by buchgr** [comment]
I have considered that too, however 1) This is executed by an application thread. There might be lots of application threads and large messages using up lots of memory that can't be freed/reused until the thread dies. 2) Android?

**Comment by carl-mastrangelo** [comment]
The memory is contained entirely in this file, and never escapes.  There should be no memory pressure added.  If there are lots of threads, the threads are going to dominate, each with multi megabyte stacks.  

For android the worst that will happen is that the buffer is allocated on the heap, and retained by the thread local.  I think this would be strictly an improvement over the current behavior on Android.

**Comment by buchgr** [comment]
We keep up to 4MiB per thread. With hundreds of application threads, there is potential to waste significant memory...

I thought Android was memory constraint? I think it could look like a memory leak. Send a few large messages and the application will suddenly need much more memory that's never reclaimed.

File: protobuf-lite/src/main/java/io/grpc/protobuf/lite/ProtoLiteUtils.java [hunk 1]
```diff
@@ -78,6 +78,18 @@ public static void setExtensionRegistry(ExtensionRegistryLite newRegistry) {
     globalRegistry = checkNotNull(newRegistry, "newRegistry");
   }
 
+  /**
+   * Local cache of buffers to use for parsing.  ThreadLocal used a WeakReference internally, so
```
Comment (buchgr) [inline protobuf-lite/src/main/java/io/grpc/protobuf/lite/ProtoLiteUtils.java]: I think a comment about thread local's implementation is not really relevant here. however it does it, that's how it does it :).

File: protobuf-lite/src/main/java/io/grpc/protobuf/lite/ProtoLiteUtils.java [hunk 2]
```diff
@@ -129,16 +141,21 @@ public T parse(InputStream stream) {
           if (stream instanceof KnownLength) {
             int size = stream.available();
             if (size > 0 && size <= GrpcUtil.DEFAULT_MAX_MESSAGE_SIZE) {
-              byte[] buf = new byte[size];
+              // Coded Input stream does not escape, so buf does not escape.
```
Comment (buchgr) [inline protobuf-lite/src/main/java/io/grpc/protobuf/lite/ProtoLiteUtils.java]: coded input stream does escape in `Parser.parse(cis)`.

File: protobuf-lite/src/main/java/io/grpc/protobuf/lite/ProtoLiteUtils.java (hunk 1 above)
Comment (ejona86) [inline protobuf-lite/src/main/java/io/grpc/protobuf/lite/ProtoLiteUtils.java]: Actually... it uses a weak ref for the thread id, but it _lazily_ cleans up the values. So these will be retained for a while even after thread death.

> To help deal with very large and long-lived usages, the hash table entries use WeakReferences for keys. However, since reference queues are not used, stale entries are guaranteed to be removed only when the table starts running out of space.

File: protobuf-lite/src/main/java/io/grpc/protobuf/lite/ProtoLiteUtils.java (hunk 2 above)
Comment (carl-mastrangelo) [inline protobuf-lite/src/main/java/io/grpc/protobuf/lite/ProtoLiteUtils.java]: Right, which is why the comment is needed here.  We assert that by the end of `parse`, there are no more references to buf (except the thread local)

File: protobuf-lite/src/main/java/io/grpc/protobuf/lite/ProtoLiteUtils.java (hunk 1 above)
Comment (carl-mastrangelo) [inline protobuf-lite/src/main/java/io/grpc/protobuf/lite/ProtoLiteUtils.java]: I originally thought that it needed to use a WeakRef, (and it isn't clear from TL). That the value will be cleaned up if under pressure. even if the thread exists from my understanding.

**Comment by carl-mastrangelo** [comment]
I'm submitting this, to see the effect on the open source benchmarks, I can revert it if the memory implications are bad enough.  I expect the number of threads to be bounded in the majority of cases.

File: protobuf-lite/src/main/java/io/grpc/protobuf/lite/ProtoLiteUtils.java (hunk 1 above)
Comment (ejona86) [inline protobuf-lite/src/main/java/io/grpc/protobuf/lite/ProtoLiteUtils.java]: I was wrong in an important way. It doesn't use WeakRef for thread id. Instead, it uses it for the ThreadLocal, which is moot. It stores the map as a field on the Thread and it is set to `null` on Thread completion, so these values could be GC'd as soon as the Thread exits.

If under memory pressure though, these values will not be GC'd, because there are strong references to them within a `Map<WeakRef<ThreadLocal>, Object>` (`ThreadLocalMap`). The only lazy cleaning is for when `ThreadLocal`s themselves are GC'd, which we don't care about.

File: protobuf-lite/src/main/java/io/grpc/protobuf/lite/ProtoLiteUtils.java (hunk 2 above)
Comment (buchgr) [inline protobuf-lite/src/main/java/io/grpc/protobuf/lite/ProtoLiteUtils.java]: but your comment says it does NOT escape, which is wrong 😅

Comment (buchgr) [inline protobuf-lite/src/main/java/io/grpc/protobuf/lite/ProtoLiteUtils.java]: No, it does escape the scope of the method. Namely to `Parser.parse(CodedInputStream)`. Whether protobuf _retains_ the byte array is a different question. I guess the terminology is off. Not important enough to do another PR though :).
